metrics/xrootd_metrics.go

```diff
@@ -1740,7 +1792,90 @@ func HandleSummaryPacket(packet []byte) error {
 	return nil
 }
 
-// handleOSSPacket processes the OSS plugin stats
+func updateIntCounter(new int, old int, counter prometheus.Counter) int {
+	incBy := math.Max(0, float64(new-old))
+	counter.Add(incBy)
+	if new < old {
+		return old
+	}
+	return new
+}
+func updateFloat64Counter(new float64, old float64, counter prometheus.Counter) float64 {
+	incBy := new - old
+	if incBy < 0 {
+		incBy = 0
+	}
+	counter.Add(incBy)
+	if new < old {
+		return old
+	}
+	return new
+}
+
+// handleS3CacheStats processes the S3 cache stats
+// It expects the blobs to be in JSON format and will update the metrics accordingly
+// It returns an error if the blobs are empty or if there is an error in unmarshalling
+// the JSON data
+// It also handles the case where the total IO or wait time is less than the previous value
+// by resetting the increment to 0.
+// The s3file_stats event schema is as follows:
+//
+//	{
+//	  "event": "s3file_stats",
+//	  "hit_b": 0,
+//	  "miss_b": 0,
+//	  "full_hit": 0,
+//	  "part_hit": 0,
+//	  "miss": 0,
+//	  "bypass_b": 12265,
+//	  "bypass": 6,
+//	  "fetch_b": 0,
+//	  "fetch": 0,
+//	  "unused_b": 0,
+//	  "prefetch_b": 0,
+//	  "prefetch": 0,
+//	  "errors": 0,
+//	  "bypass_s": 1.27,
+//	  "fetch_s": 0
+//	}
+func handleS3CacheStats(blobs [][]byte) error {
+
+	// we need to process the blobs backwards to ensure that we are processing the last valid event
+	// the most relevant data is the last valid event in the sequence of blobs
+	for i := len(blobs) - 1; i >= 0; i-- {
+		blob := blobs[i]
+		s3fileStats := OssS3CacheGs{}
+		if err := json.Unmarshal(blob, &s3fileStats); err != nil {
+			return errors.Wrap(err, "failed to parse S3 file stats json")
+		}
+		if !allowedEvents[s3fileStats.Event] {
+			log.Debugf("handleS3CacheStats received an S3 file stats packet with an unrecognized event type (%s)", s3fileStats.Event)
```

Comment:
If this is something we really don't expect to happen, it may be better to set this at the warning or error level so it's a bit more obvious. I'll defer to you.